File: javascript/packages/linter/src/rules/html-no-abstract-roles.ts
import { ParserRule } from "../types.js"
import { AttributeVisitorMixin, ABSTRACT_ARIA_ROLES, StaticAttributeStaticValueParams } from "./rule-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { ParseResult } from "@herb-tools/core"

class NoAbstractRolesVisitor extends AttributeVisitorMixin {
  protected checkStaticAttributeStaticValue({ attributeName, attributeValue, attributeNode }: StaticAttributeStaticValueParams): void {
    if (attributeName !== "role") return
    if (!attributeValue) return

    const normalizedValue = attributeValue.toLowerCase()

    if (!ABSTRACT_ARIA_ROLES.has(normalizedValue)) return

    this.addOffense(
      `The \`role\` attribute must not use abstract ARIA role \`${attributeValue}\`. Abstract roles are not meant to be used directly.`,
      attributeNode.location,
    )
  }
}

export class HTMLNoAbstractRolesRule extends ParserRule {
  name = "html-no-abstract-roles"

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoAbstractRolesVisitor(this.name, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_no_abstract_roles.rs
use crate::utils::aria_data::ABSTRACT_ARIA_ROLES;
use crate::utils::tag_utils::{get_attribute_in, get_static_attribute_value};

use herb::nodes::{AnyNode, ERBOpenTagNode, HTMLOpenTagNode};
use herb::Visitor;

rule_visitor!(NoAbstractRolesVisitor);
define_parser_rule!(HTMLNoAbstractRolesRule, "html-no-abstract-roles", Warning, NoAbstractRolesVisitor,
  parser_options: { action_view_helpers: true },
  introduced_in: "0.9.0"
);

impl NoAbstractRolesVisitor {
  fn check_children(&mut self, children: &[AnyNode]) {
    if let Some(role_attribute) = get_attribute_in(children, "role") {
      if let Some(role_value) = get_static_attribute_value(role_attribute) {
        if !role_value.is_empty() {
          let normalized_value = role_value.to_lowercase();

          if ABSTRACT_ARIA_ROLES.contains(normalized_value.as_str()) {
            self.add_offense(
              format!(
                "The `role` attribute must not use abstract ARIA role `{}`. Abstract roles are not meant to be used directly.",
                role_value
              ),
              role_attribute.location.clone(),
            );
          }
        }
      }
    }
  }
}

impl Visitor for NoAbstractRolesVisitor {
  fn visit_html_open_tag_node(&mut self, node: &HTMLOpenTagNode) {
    self.check_children(&node.children);
    self.walk_html_open_tag_node(node);
  }

  fn visit_erb_open_tag_node(&mut self, node: &ERBOpenTagNode) {
    self.check_children(&node.children);
    self.walk_erb_open_tag_node(node);
  }
}
